### DISCUSSION TITLE
Server action failed to return when url search params changed during fetching

### DISCUSSION THREAD

**Discussion by soranoo** [vercel/next.js#69167] [Help]
### Link to the code that reproduces this issue

https://github.com/soranoo/nextjs-server-action-failed-to-return-when-url-changed-during-fetching

### To Reproduce

1. Clone the rp
2. run `npm i`
3. Goto `http://127.0.0.1:3000`
4. Press the `Set Params during fetching` button
![image](https://github.com/user-attachments/assets/2b663605-0a1f-4d8a-94ba-320965273d3f)

### Current vs. Expected behavior

Current: The URL search param is set but no return from server action (keep fetching)

Expected: The URL search param is set and a message from server action gets displayed

### Provide environment information

```bash
Operating System:
  Platform: win32
  Arch: x64
  Version: Windows 11 Pro
  Available memory (MB): 30414
  Available CPU cores: 20
Binaries:
  Node: 20.11.1
  npm: N/A
  Yarn: N/A
  pnpm: N/A
Relevant Packages:
  next: 15.0.0-canary.108 // Latest available version is detected (15.0.0-canary.108).
  eslint-config-next: N/A
  react: 19.0.0-rc-e948a5ac-20240807
  react-dom: 18.3.1
  typescript: 5.5.4
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

create-next-app, Font (next/font), Navigation

### Which stage(s) are affected? (Select all that apply)

next dev (local), next build (local), next start (local), Vercel (Deployed), Other (Deployed)

### Additional context

- Tested in `14.2.5` - Failed
- Tested in `15.0.0-canary.108` - Failed

**Comment by samcx** [comment]
@soranoo 

I inputed all of this into `v0`.

The issue causing `isFetching` to stay true indefinitely is likely due to a race condition between state updates and the refetch operation. Here's a breakdown of what's happening:

1. When the `setParamsAfterRequest` function is called:

1. It resets the URL parameters.
2. It updates the `msg` state.
3. It triggers a refetch of the data.



2. The `useEffect` hook is dependent on the `msg` state:


```javascript
useEffect(() => {
  setParams();
}, [msg]);
```

3. When `msg` is updated, it triggers this effect, which calls `setParams()`.
4. `setParams()` updates the URL, which causes a re-render of the component.


This creates a loop:

1. Button click → Update state → Refetch
2. State update → Effect runs → URL update
3. URL update → Component re-render → Refetch (potentially)


The constant re-renders and potential refetches can interfere with the query's ability to complete, keeping `isFetching` true.

To fix this issue, you can:

1. Remove the `useEffect` hook that depends on `msg`.
2. Update the `setParamsAfterRequest` function to handle all operations in the correct order.


Here's a corrected version of the `setParamsAfterRequest` function:

```typescript
const setParamsAfterRequest = async () => {
  setMsg(`Params set during fetching ${new Date().toISOString()}`);
  await refetch();
  router.replace(`?t="hello"`, { scroll: false });
};
```

This change ensures that:

1. The message is updated.
2. The data is refetched and the operation completes.
3. The URL is updated only after the refetch is complete.


Additionally, you should remove the `useEffect` hook that depends on `msg`:

```typescript
// Remove this effect
// useEffect(() => {
//   setParams();
// }, [msg]);
```

These changes should prevent the infinite fetching issue while still maintaining the desired functionality. The URL will be updated after each successful data fetch, and the `isFetching` state should correctly reflect the actual fetching status .

**Reply by samcx** [reply]
@soranoo Can you clarify exactly what you mean by used by two different components?

As far as I can see, all your logic is in the `Home` component.

**Reply by soranoo** [reply]
[the example repo was updated]

I have to apologise again since I used an incorrect description in the previous reply (it has been deleted to avoid any misleading to other readers).

I got a post site and the post will set its title as a search parameter for better understanding when the user shares the site with others.
eg. `https;//example.com/posts/abcd?post-title="im-a-good-post"` instead of `https;//example.com/posts/abcd`

#### Setup
1. The post's title will be set once the related component gets mounted  (`SeactionA` in my example)
2. Some data will be fetched once the related component gets mounted  (`SeactionB` in my example)

#### Page structure
```
Page
|-Section A (set search param)
|-Section B (fetch data)
|
```
(both section A and B will be mounted at the same time)

#### How to use the example
1. Start the dev server
2. visit the localhost (the title will be set and the fetch will never return (no msg show))

Please feel free to ask for clarification, I know I am terrible at describing things...

**Reply by samcx** [reply]
@soranoo Can you also describe what you see and what you expect to see (e.g., a short video)?

**Reply by soranoo** [reply]
Current
![image](https://github.com/user-attachments/assets/5c0d15fa-42cb-4d81-8360-0e5ac8198d3b)
Expected
![image](https://github.com/user-attachments/assets/f475671e-2a03-4216-ba74-556e087e6e5d)
